File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/update/UpdateRequest.java
package fr.myprysm.vertx.elasticsearch.action.update;

import fr.myprysm.vertx.elasticsearch.action.support.DocWriteRequest;
import fr.myprysm.vertx.elasticsearch.action.support.FetchSourceContext;
import fr.myprysm.vertx.elasticsearch.action.support.Script;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.experimental.Accessors;
import org.elasticsearch.action.support.WriteRequest;
import org.elasticsearch.index.VersionType;

import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
@DataObject(generateConverter = true)
public class UpdateRequest extends DocWriteRequest {
    private JsonObject doc;
    private JsonObject upsert;

    private Script script;

    private List<String> fields;
    private FetchSourceContext fetchSourceContext;

    private int retryOnConflict = 0;

    private Integer waitForActiveShards;

    private boolean docAsUpsert = false;
    private boolean scriptedUpsert = false;
    private boolean detectNoop = true;

    public UpdateRequest(String index, String type, String id) {
        super(index, type, id, org.elasticsearch.action.DocWriteRequest.OpType.UPDATE);
    }

    public UpdateRequest(UpdateRequest other) {
        super(other);
        doc = other.doc;
        upsert = other.upsert;
        script = other.script;
        fields = other.fields;
        fetchSourceContext = other.fetchSourceContext;
        retryOnConflict = other.retryOnConflict;
        waitForActiveShards = other.waitForActiveShards;
        docAsUpsert = other.docAsUpsert;
        scriptedUpsert = other.scriptedUpsert;
        detectNoop = other.detectNoop;
    }

    public UpdateRequest(JsonObject json) {
        super(json);
        UpdateRequestConverter.fromJson(json, this);
    }

    public JsonObject toJson() {
        JsonObject json = super.toJson();
        UpdateRequestConverter.toJson(this, json);
        return json;
    }

    @Override
    public UpdateRequest setIndex(String index) {
        return (UpdateRequest) super.setIndex(index);
    }

    @Override
    public UpdateRequest setType(String type) {
        return (UpdateRequest) super.setType(type);
    }

    @Override
    public UpdateRequest setId(String id) {
        return (UpdateRequest) super.setId(id);
    }

    @Override
    public UpdateRequest setRouting(String routing) {
        return (UpdateRequest) super.setRouting(routing);
    }

    @Override
    public UpdateRequest setParent(String parent) {
        return (UpdateRequest) super.setParent(parent);
    }

    @Override
    public UpdateRequest setVersion(long version) {
        return (UpdateRequest) super.setVersion(version);
    }

    @Override
    public UpdateRequest setVersionType(VersionType versionType) {
        return (UpdateRequest) super.setVersionType(versionType);
    }

    @Override
    public UpdateRequest setOpType(org.elasticsearch.action.DocWriteRequest.OpType opType) {
        return (UpdateRequest) super.setOpType(opType);
    }

    @Override
    public UpdateRequest setRefreshPolicy(WriteRequest.RefreshPolicy refreshPolicy) {
        return (UpdateRequest) super.setRefreshPolicy(refreshPolicy);
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/update/UpdateRequest.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.update

import fr.myprysm.vertx.elasticsearch.action.update.UpdateRequest
import fr.myprysm.vertx.elasticsearch.action.support.FetchSourceContext
import fr.myprysm.vertx.elasticsearch.action.support.Script
import org.elasticsearch.action.DocWriteRequest.OpType
import org.elasticsearch.action.support.WriteRequest.RefreshPolicy
import org.elasticsearch.index.VersionType

fun UpdateRequest(
  detectNoop: Boolean? = null,
  doc: io.vertx.core.json.JsonObject? = null,
  docAsUpsert: Boolean? = null,
  fetchSourceContext: fr.myprysm.vertx.elasticsearch.action.support.FetchSourceContext? = null,
  fields: Iterable<String>? = null,
  headers: Map<String, String>? = null,
  id: String? = null,
  index: String? = null,
  opType: OpType? = null,
  parent: String? = null,
  refreshPolicy: RefreshPolicy? = null,
  retryOnConflict: Int? = null,
  routing: String? = null,
  script: fr.myprysm.vertx.elasticsearch.action.support.Script? = null,
  scriptedUpsert: Boolean? = null,
  type: String? = null,
  upsert: io.vertx.core.json.JsonObject? = null,
  version: Long? = null,
  versionType: VersionType? = null,
  waitForActiveShards: Int? = null): UpdateRequest = fr.myprysm.vertx.elasticsearch.action.update.UpdateRequest().apply {

  if (detectNoop != null) {
    this.setDetectNoop(detectNoop)
  }
  if (doc != null) {
    this.setDoc(doc)
  }
  if (docAsUpsert != null) {
    this.setDocAsUpsert(docAsUpsert)
  }
  if (fetchSourceContext != null) {
    this.setFetchSourceContext(fetchSourceContext)
  }
  if (fields != null) {
    this.setFields(fields.toList())
  }
  if (headers != null) {
    this.setHeaders(headers)
  }
  if (id != null) {
    this.setId(id)
  }
  if (index != null) {
    this.setIndex(index)
  }
  if (opType != null) {
    this.setOpType(opType)
  }
  if (parent != null) {
    this.setParent(parent)
  }
  if (refreshPolicy != null) {
    this.setRefreshPolicy(refreshPolicy)
  }
  if (retryOnConflict != null) {
    this.setRetryOnConflict(retryOnConflict)
  }
  if (routing != null) {
    this.setRouting(routing)
  }
  if (script != null) {
    this.setScript(script)
  }
  if (scriptedUpsert != null) {
    this.setScriptedUpsert(scriptedUpsert)
  }
  if (type != null) {
    this.setType(type)
  }
  if (upsert != null) {
    this.setUpsert(upsert)
  }
  if (version != null) {
    this.setVersion(version)
  }
  if (versionType != null) {
    this.setVersionType(versionType)
  }
  if (waitForActiveShards != null) {
    this.setWaitForActiveShards(waitForActiveShards)
  }
}

